### ISSUE TITLE
feature request - get code multi

### ISSUE THREAD

**Issue by ben-polytope** [mds1/multicall#60]
Hi, thank you for the development of multicall3, it is very useful.

I have a request for a feature that can further reduce node calls. Rather than hitting the eth rpc method eth_getCode, we can instead implement a getCodeMulti method, something like the below:

```
    struct ReturnCode {
         address addr;
         bytes code;
     } 

    function getCodeMulti(address[] calldata addresses) public payable returns (ReturnCode[] memory returnContracts) {
         uint256 length = addresses.length;
         returnContracts = new ReturnCode[](length);
         address addr;
         for (uint256 i=0; i < length;) {
             addr = addresses[i];
             returnContracts[i] = ReturnCode(addr, addr.code);
             unchecked {++ i;}
         }
     }
 ```
 
 This function would allow for users to get bytecode from multiple contract addresses with only one rpc call. This may be out of scope for this project as it would require redeployment accross all networks, but happy to open a PR or contribute where I can if this feature seems like something worth adding.

**Comment by mds1** [comment]
Ah interesting, you could also do something similar with code hashes too. I wouldn't count on a new Multicall deploy anytime soon just to add this, so if it's something you currently need you should consider deploying your own version.

One thing worth noting is that if you're returning a lot of bytecode, it all needs to be copied to memory. Gas cost increases quadratically with memory expansion, so you'll very quickly use a lot of gas to execute this. Most RPCs (like Infura/Alchemy) have an `eth_call` gas limit of ~10x the block gas limit IIRC, to prevent DoS'ing them, though of course you can have no gas limit if running against a local node.

You can also consider batch JSON RPC requests: https://geth.ethereum.org/docs/interacting-with-geth/rpc/batch

**Comment by mds1** [comment]
Closing this per above comment since it won't be implemented for the current version of Multicall 🙂

**Comment by timaiv** [comment]
@mds1 @ben-polytope 

I'm also interested in similar functionality, but in particalur for: `isContract`.

```
function isContract(address _addr) private returns (bool isContract){
  uint32 size;
  assembly {
    size := extcodesize(_addr)
  }
  return (size > 0);
}
```

**Comment by Pybast** [comment]
@timaiv I found a contract that has such functionality.

For others who might be looking for contracts with other public / external functions I will share below the method I used to find the contract.

I used search operators on google, which allowed me to get down to only a few good candidates and after a little manual checking of the source code, I was able to find one that I can use.
The google search I used is the following 

`"function isContract(address account) public view returns (bool)" site:etherscan.io`

Here is the contract that implements the `isContract` function 

[0xD17C3BD8b499F0965181473596098dE3BE905B0e](https://etherscan.io/address/0xD17C3BD8b499F0965181473596098dE3BE905B0e)
